tools/src/main/java/org/neo4j/tools/CommandReactor.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.tools;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.PrintStream;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.neo4j.helpers.Args;
+
+/**
+ * Useful utility which listens to input from console and reacts to each line, i.e. after each ENTER.
+ * Reactions are added with {@link #add(String, Action)} and are then reacted to in a separate thread,
+ * continuously listening to console input.
+ */
+public class CommandReactor
```

Comment:
The order of members in here is less than ideal.

I'd like for the `Action` interface to be defined very early, since it is an interface users of the class needs to implement. Then I'd like for the public methods to come before the private implementation, i.e. the inner `Reactor` class should be last.